### ISSUE TITLE
Customizable Schema Validation Error Messages

### ISSUE THREAD

**Issue by jinhong-** [gregsdennis/Manatee.Json#161]
Any chance this can be supported? The only way i can think of now is to use some form of string pattern matching

**Comment by gregsdennis** [comment]
I like the idea. Can you describe what you want or give an example?

**Comment by jinhong-** [comment]
To elaborate a little more:
I think the project right now will need some tweaks to support this feature. We would first need to introduce some sort of extensibility point, perhaps some sort of pipeline or interceptor or decorator per property validator

What i meant by pattern matching was actually to just get the validation results, and then match it with the expected string value 😄

**Comment by gregsdennis** [comment]
I can manage the implementation, but what I need from you is an example of how you'd expect to use the feature.

**Comment by jinhong-** [comment]
I am honestly not sure yet. But I would need to be able to tweak the messages during runtime for localization
I can think of a few ways i could use it
1. Expose some custom error message format property per IJsonSchemaPropertyValidator, so that I can set the error message for each error
2. Have some sort of Error message map or IErrorMessageFormatter interface which is used to handle all error messages, but it should be able to distinguish the different validation error types. Minimum, Maximum, Required, etc. Also, it will need to cater for custom errors because there might be Custom validators
3. SchemaValidationResults will not just return property and message, but will also return some sort of error type, which would allow for better disambiguation between different validation error types. But this will also mean it would need a way for the validator to report the expected and actual value as well

But ultimately, the idea is to be able to intercept the error message value and replace it with something else. It would be a bonus to print the configured validation value as well. For eg

A max length validator would say "Expected: 2; Actual: 5 but i might want to specify a custom string with format values saying "{propertyName} is {actual} characters long, which exceeds length of {expected}"

**Comment by gregsdennis** [comment]
Okay. I see what you're after. I'll see what mechanism I can come up with. May be as simple as a static registry class.

**Comment by gregsdennis** [comment]
I think a static registry is probably the best mechanism for this.  It would store error message templates.  I think tokenizing the messages would be more explicit than simple format strings. It would allow the author to know exactly what parameters go where.  Either way, I'll need to document it.

To address your suggestions:

1. I don't exactly have a single validator for every property. This is because many of the properties are related, like `properties`, `additionalProperties`, and `patternProperties`. These all have a single validator, so it outputs multiple messages.
2. I think this is the closest to what I suggest.
3. Adding an error code is probably a good idea, too. I'll try to incorporate that.

**Comment by jinhong-** [comment]
Cool that was quick!
One question. I noticed that the error messages do not really include the property name, for eg
https://github.com/gregsdennis/Manatee.Json/blob/f754f28983e8c5e449f0a7c4f4a8073679b28d43/Manatee.Json/Schema/Validators/MaximumSchemaPropertyValidator.cs#L22
Is there any particular reason why?

**Comment by gregsdennis** [comment]
Yes, the property name is added as the error bubbles up through the `properties` validator. I want to say I had a constructor overload conflict if I had just tried to pass the error message, but I don't recall.

**Comment by jinhong-** [comment]
Ah! Sorry i didn't catch that part. So ultimately "property" will be resolved

**Comment by jinhong-** [comment]
I think it is still not sufficient for what I am trying to do. I am still having difficulties on the ability to customize error messages. This was supposed to be part of the specifications but got dropped: https://github.com/json-schema-org/json-schema-spec/issues/62

I don't think the library needs to implement this but would it make sense to include 2 things in the SchemaValidationResults:
1. The current schema that is validated, so that I can extract any custom metadata attached to the schema
2. Inner/nested errors emitted by composite validators like AllOf, OneOf, AnyOf, If. There are 2 ways to emit this. One is to include it as part of the error list of SchemaValidationResults, or create some sort of InnerResults property

**Comment by jinhong-** [comment]
Consider this example:
{
  "allOf": [
    {
      "anyOf": [
        {
          "required": [
            "firstName"
          ]
        },
        {
          "required": [
            "lastName"
          ]
        },
        {
          "required": [
            "middleName"
          ]
        }
      ]
    },
    {
      "anyOf": [
        {
          "required": [
            "mobileNumber"
          ]
        },
        {
          "required": [
            "email"
          ]
        }
      ]
    }
  ]
}

Here i would like to verify if firstName or middleName or lastName is present, and i would also like to verify if mobileNumber or email is present. 
When validation error happens, I would like to be able to customize to "mobileNumber or email is required". At the moment i believe it will print 2 error messages, saying mobileNumber is required, email is required. This is untrue because only 1 value is required. Not both.

Maybe rather than flattening out the error messages into SchemaValidationResults, it might be good to maintain the actual tree. For example i could do something like this:

{
  "allOf": [
    {
      "anyOf": [
        {
          "required": [
            "firstName"
          ]
        },
        {
          "required": [
            "lastName"
          ]
        },
        {
          "required": [
            "middleName"
          ]
        }
      ],
      "manatee:customErrors":{
        "anyOf": "firstName, lastName, middleName is required"
      }
    },
    {
      "anyOf": [
        {
          "required": [
            "mobileNumber"
          ]
        },
        {
          "required": [
            "email"
          ]
        }
      ],
      "manatee:customErrors":{
        "anyOf": "mobileNumber or email is required"
      }
    }
  ]
}

**Comment by gregsdennis** [comment]
@jinhong- What you're proposing would be a significant departure from how I handle errors.  It would have to know that all of these are the same error type and combine them in a human-friendly way.

That said, I take your point about knowing more information on what subschema is raising the error.  Currently, only the JSON instance (the only being validated) is reported on.

You may be keen to look at https://github.com/json-schema-org/json-schema-spec/issues/396.  This is a proposal which formalizes the output of JSON Schema for all implementations.

I think there are some changes that could be made, but probably won't be until the next major version since they're likely breaking.

**Comment by jinhong-** [comment]
I tried to do some small experiments. I think the change doesn't need to be breaking from API perspective but there will be behavioral changes. Check my fork:
https://github.com/jinhong-/Manatee.Json/blob/master/Manatee.Json/Schema/SchemaValidationError.cs

I added 3 properties i believe will be sufficient to solve what i need

**Comment by jinhong-** [comment]
Including these pieces of information would greatly enhance what can be extracted. I could do some post processing since i have the schema, plus, i can also add metadata into the schema as well. for eg

```
{
  "allOf": [
    {
      "anyOf": [
        {
          "required": [
            "firstName"
          ]
        },
        {
          "required": [
            "lastName"
          ]
        },
        {
          "required": [
            "middleName"
          ]
        }
      ],
      "manatee:customErrors":{
        "anyOf": "firstName, lastName, middleName is required"
      },
	  **"custom:logErrorAs": "error"**
    },
    {
      "anyOf": [
        {
          "required": [
            "mobileNumber"
          ]
        },
        {
          "required": [
            "email"
          ]
        }
      ],
      "manatee:customErrors":{
        "anyOf": "mobileNumber or email is required"
      },
	  **"custom:logErrorAs": "warning"**
    }
  ]
}
```

**Comment by gregsdennis** [comment]
@jinhong- a couple things:

First, please have a look at https://github.com/json-schema-org/json-schema-spec/issues/396 and see if this will satisfy what you're looking for.

Secondly, until that is ratified, do you think it'd be helpful to have, in addition to the current error string, an error data object that has an error code, expected data, and actual data so that you can build your own messages from that? I think I can work that up.

**Comment by freakingawesome** [comment]
Is there any current documentation on customizing error messages? It looks like [`SchemaErrorMessages`](https://github.com/gregsdennis/Manatee.Json/blob/master/Manatee.Json/Schema/SchemaErrorMessages.cs) is no longer referenced in the project and I cannot figure out how to create readable error messages.

Thanks for the great library!

**Comment by gregsdennis** [comment]
@freakingawesome, I haven't done any docs on it yet. I actually haven't done any docs on the output formatting at all really because I'm waiting for draft-08 to finalize lest anything really change.

You're welcome to review the Output Formatting section of the [spec draft](https://github.com/json-schema-org/json-schema-spec/), though, for a bit of a primer.  I'm also happy to answer any questions beyond that, but it's a lot to try and summarize here.

Outside of what the spec says, I also include a `keyword` property as well as an `AdditionalInformation` property in each node that can be used to formulate your messages.
